docs/security.md

```diff
@@ -0,0 +1,187 @@
+
+So, for example, a dapp is not allowed to call the `keyring_submitRequest`
+method of your Snap, and MetaMask is not allowed to call the
+`keyring_createAccount` method.
+
+MetaMask also enforces that only dapps allowlisted in the Snap's manifest can
+call the methods above. For example, if your Snap's manifest contains the
+following allowlist:
+
+```jsonc
+{
+  // Other fields of the manifest...
+  "initialPermissions": {
+    // Other initial permissions...
+    "endowment:keyring": {
+      "allowedOrigins": ["https://<dapp domain>"]
+    }
+  }
+}
+```
+
+Then only the dapp at `https://<dapp domain>` can call the methods above. If a
+dapp hosted in a different domain attempts to call one of these methods, the
+request will be rejected by MetaMask.
+
+But Snap developers are advised to further constrain the methods that are
+exposed to dapps in accordance with their Snap's functionality. For instance,
+if your Snap does not support account deletion via dapps, your Snap should
+reject calls to the `keyring_deleteAccount` method originating from dapps.
+
+Your Snap can also impose varying restrictions depending on the calling dapp.
+For example, _dapp-1_ may have access to a different set of methods than
+_dapp-2_.
+
+The following code snippet provides an example of how to implement such logic:
+
+```ts
+const permissions: Record<string, string[]> = {
+  'https://<dapp-1 domain>': [
+    // List of allowed methods for dapp-1.
+  ],
+  'https://<dapp-2 domain>': [
+    // List of allowed methods for dapp-2.
+  ],
+};
+
+if (origin !== 'metamask' && !permissions[origin]?.includes(request.method)) {
+  // Reject the request.
+}
+```
+
+Notice, however, that both dapps must be allowlisted in the Snap's manifest.
+
+### Ensure that the redirect URL cannot be manipulated
+
+If your Snap implements the [asynchronous transaction
+flow](./architecture.md#transaction-flow), ensure that the redirect URL is
+valid and cannot be manipulated, otherwise the user could be redirected to a
+malicious website.
+
+```ts
+async submitRequest(request: KeyringRequest): Promise<SubmitRequestResponse> {
+  // Your Snap's custom logic here...
+  return {
+    pending: true,
+    redirect: {
+      message: 'Please continue in the Dapp.',
+      url: 'https://<dapp domain>/sign?tx=1234', // !!! ENSURE THIS IS A SAFE URL !!!
+    },
+  };
+}
+```
+
+> [!IMPORTANT]
+> Only HTTPS URLs are allowed in the `url` field, and the provided URL will be
+> checked against a [list of blocked domains][eth-phishing-detect]. However,
+> for development purposes, HTTP URLs are allowed on Flask.
+
+### Remove all debug code from your production Snap
+
+Ensure that all debug code is removed from your production Snap. This mistake
+can lead to multiple security vulnerabilities. For example, secret information
+may be logged to the console, or a security check may be bypassed by a
+malicious dapp.
+
+### Don't expose Keyring methods through the `onRpcRequest` export
```

Comment:
TODO: check if we can check which handler the message is coming from

Reply:
Not sure if it's possible since the handler type isn't passed to the handler function.